encoding/template/run_length_decode.go

```diff
@@ -0,0 +1,72 @@
+package template
+
+import (
+	"encoding/binary"
+
+	"github.com/xichen2020/eventdb/x/io"
+)
+
+// WriteValueFn reads a GenericValue from an `io.Reader`.
+type WriteValueFn func(reader io.Reader) (GenericValue, error)
+
+// runLengthDecodeValue run length decodes a stream of GenericValues.
+func runLengthDecodeValue(
+	reader io.Reader,
+	writeValue WriteValueFn,
+) *RunLengthValueIterator {
+	return newRunLengthValueIterator(reader, writeValue)
+}
+
+// RunLengthValueIterator iterates over a run length encoded stream of GenericValue data.
+type RunLengthValueIterator struct {
+	reader      io.Reader
+	writeValue  WriteValueFn
+	curr        GenericValue
+	repetitions int64
+	closed      bool
+	err         error
+}
+
+// Next iteration.
+func (rl *RunLengthValueIterator) Next() bool {
+	if rl.closed || rl.err != nil {
+		return false
+	}
+
+	if rl.repetitions > 1 {
```

Comment:
Hm should this be `rl.repetitions >= 1`? If so can you add a unit test for this (assuming there's one)?

Reply:
The unit test case I have covers this. Since the logic was changed to: "if a value shows up once, the # of repetitions starts at 1." So what would happen is:
- First pass, # of repetitions is 0.
- Read in value, # of repetitions is 1.
- Read in next value, # of repetitions is 2.
- Loop once based on this condition and decrement # of repetitions.